File: api/src/main/java/rt/bugtrack/core/bug/entity/Bug.java
package rt.bugtrack.core.bug.entity;

import java.util.ArrayList;
import java.util.List;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import rt.bugtrack.auth.entity.User;
import rt.bugtrack.core.bug.util.Status;

@Entity
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "bug")
public class Bug {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Integer id;

    @Column(name = "project_id")
    private Integer projectId;

    @ManyToOne
    @JoinColumn(name = "reporter_id", referencedColumnName = "id")
    private User reporter;

    @Builder.Default
    @OneToMany(mappedBy = "bug", cascade = CascadeType.ALL)
    private List<Step> steps = new ArrayList<>();

    @Enumerated(EnumType.STRING)
    @Column(name = "status")
    private Status status;

    @Column(name = "title")
    private String title;

    @Column(name = "description")
    private String description;
}

File: ui/src/lib/core/types/Bug.ts
import type { Status } from "./Status"
import type { Step } from "./Step"
import type { User } from "./User"

export type Bug = {
    id?: number,
    projectId?: number,
    reporter?: User,
    steps?: Step[],
    title?: string,
    description?: string,
    status?: Status,
    editable?: boolean
}

export type BugHistory = {
    id?: number,
    user?: User,
    bugId?: User,
    status?: Step[],
    title?: string,
    description?: string,
    stepsUpdated?: boolean,
    changeTime?: number,
    changeDateTime?: Date,
    changedFields?: string[],
    isCreated?: boolean
}
